### DISCUSSION TITLE
First feedback for v2

### DISCUSSION THREAD

**Discussion by jurgenhaas** [cweagans/composer-patches#484] [General]
Found a couple of issues when giving `dev-main` a first try:

- [Configuration error](https://docs.cweagans.net/composer-patches/usage/configuration/): Neither patches nor the patches-file should go into `extras/composer-patches`, they need to go into extras directly. Otherwise, they don't do anything.
- Remote patches-file doesn't work anymore, e.g. we maintain a common patch-file for all our Drupal projects at https://gitlab.lakedrops.com/composer/plugin/drupal-environment/-/raw/develop/patches/d10.json, but this is no longer used with v2
- If a given patches-file file doesn't exist, there should be a warning on the console
- When a patch can not be applied, composer fails and stops. In v1, it ignored that and just continued. Has that changed deliberately?
- Workflow is a bit unclear: in v1, updates to the patch list was recognized automatically and applied when we ran `composer update`. With the new patches.lock file, do we have to force the repatching now manually, i.e. first relock the patch file and then repatch?

**Comment by cweagans** [comment]
Thank you for the feedback!

> [Configuration error](https://docs.cweagans.net/composer-patches/usage/configuration/): Neither patches nor the patches-file should go into extras/composer-patches, they need to go into extras directly. Otherwise, they don't do anything.

https://github.com/cweagans/composer-patches/pull/494 addresses the `patches-file` issue (so `extra/composer-patches/patches-file` is now the right location). `patches` should still be defined directly in `extra` (so `extra/patches`).

> Remote patches-file doesn't work anymore, e.g. we maintain a common patch-file for all our Drupal projects at https://gitlab.lakedrops.com/composer/plugin/drupal-environment/-/raw/develop/patches/d10.json, but this is no longer used with v2

Opened https://github.com/cweagans/composer-patches/issues/495

> If a given patches-file file doesn't exist, there should be a warning on the console

Not sure I agree here. The default patches file in the new plugin is `patches.json`, which will work without specifying anything in the configuration. If it's not there, wouldn't it be better to resolve patches from the other available resolvers?

> When a patch can not be applied, composer fails and stops. In v1, it ignored that and just continued. Has that changed deliberately?

This is intentional. If a user specifies that a patch should be applied, I'm not sure it's valid to just ignore when that patch can't be applied.

> Workflow is a bit unclear: in v1, updates to the patch list was recognized automatically and applied when we ran composer update. With the new patches.lock file, do we have to force the repatching now manually, i.e. first relock the patch file and then repatch?

https://docs.cweagans.net/composer-patches/usage/recommended-workflows/#add-a-patch-to-your-project

Yep. In 2.x, you have to be a bit more explicit sometimes. In 1.x, people were having problems with the plugin just randomly deleting directories full of code (some of which may have been work-in-progress).

**Comment by jurgenhaas** [comment]
>>If a given patches-file file doesn't exist, there should be a warning on the console

>Not sure I agree here. The default patches file in the new plugin is patches.json, which will work without specifying anything in the configuration.

Hmm, that doesn't sound right. Imagine, someone has a typo in their configuration and won't notice for months, just because nothing ever complained about it? I see your point re. the default value, but that could be handled by not complaining if the value equals the default and that isn't there. In all other cases, a missing file should trigger an error.

>>When a patch can not be applied, composer fails and stops. In v1, it ignored that and just continued. Has that changed deliberately?

>This is intentional. If a user specifies that a patch should be applied, I'm not sure it's valid to just ignore when that patch can't be applied.

It was convenient in the past, as such "errors" most often occurred when the patch was merged by the package maintainer. In that case, we saw the list of not applying patches and then reviewed them all at once to correct the complete list of patches. With the new model, which makes perfect sense in an ideal world, we will only get a chance to notice the first failing patch, fix that and then run the update again. That has to go on until all of them have been identified.

In the past, wasn't there a command line argument to switch to this strict mode as well? If so, wouldn't it make sense to no have such a switch with the reverse functionality, i.e. fail by default but allow to overwrite that e.g. in dev environments?

>Workflow ...

That makes sense, we've got to adopt to that behaviour and I guess the change will then have to be even bigger, especially in deployment pipelines. I guess we then need to go as far as this:

- Run composer install without executing scripts or plugins
- Then apply patches
- Afterwards, trigger all the plugin hooks and scripts

Not sure, if that's even possible !?

**Comment by cweagans** [comment]
>  I see your point re. the default value, but that could be handled by not complaining if the value equals the default and that isn't there. In all other cases, a missing file should trigger an error.

I'm open to putting a warning in the output if the patches file is missing.

> That has to go on until all of them have been identified.

This is what you have to do with Composer if e.g. you have an incompatible set of packages or you're missing platform deps. I'm open to a switch to make it not fail.

>  I guess we then need to go as far as this [...]

I don't understand why you would need to do that. Just `composer install` as normal and patches will be applied. If you're changing patches, there is a bit more work to do, but it shouldn't change that much. Also, `composer patches-repatch` should trigger all of the normal composer hooks (it's actually running `composer install` under the hood), so your normal scripts and such should work?

**Comment by mxr576** [comment]
@cweagans Hi, 
can you share an update about your v2 plans? I believe your priorities have changed because several months have passed without public activity since you asked feedback about v2. Is there an ETA for a stable 2.0.0 release?

**Reply by cweagans** [reply]
https://www.cweagans.net/2023/07/dependency-patch-resolution/

Next steps are further testing, bug fixes, etc. I do not want to hold up the release any longer than I have to.
